tests/python/relay/transform/test_compiler_function_utils.py

```diff
@@ -0,0 +1,162 @@
+                     PartitionedFromPattern="nn.batch_matmul_", Composite="cublas.batch_matmul") -> Tensor[(600, 32, 32), float16] {
+              nn.batch_matmul(%FunctionVar_0_01, %FunctionVar_0_11, out_dtype="float16", transpose_b=True)
+            };
+            %6(%y_3_i0, %y_3_i1)
+          };
+          %3 = %2(%x3, meta[relay.Constant][2]);
+          (%1, %3)
+        }
+        
+        def @tvmgen_default_cutlass_main_0(%y_0_i0: Tensor[(1600, 768), float16], %y_0_i1: Tensor[(2304, 768), float16], %y_0_i2: Tensor[(2304), float16],
+                  Inline=1, Compiler="cutlass", global_symbol="tvmgen_default_cutlass_main_0", Primitive=1) -> Tensor[(1600, 2304), float16] {
+          %4 = fn (%FunctionVar_0_0: Tensor[(1600, 768), float16], %FunctionVar_0_1: Tensor[(2304, 768), float16], %FunctionVar_0_2: Tensor[(2304), float16],
+                   PartitionedFromPattern="nn.dense_add_", Composite="cutlass.dense_bias") -> Tensor[(1600, 2304), float16] {
+            %5 = nn.dense(%FunctionVar_0_0, %FunctionVar_0_1, units=2304);
+            add(%5, %FunctionVar_0_2)
+          };
+          %4(%y_0_i0, %y_0_i1, %y_0_i2)
+        }
+        """,
+        "from_string",
+        None,
+        metatable,
+    )
+
+
+def expected_extern_mod():
+    return tvm.parser.parse(
+        """
+        #[version = "0.0.5"]
+        def @main(%x0 : Tensor[(1600, 768), float16], %x3 : Tensor[(600, 32, 64), float16]) -> (Tensor[(1600, 2304), float16], Tensor[(600, 32, 32), float16]) {
+          %1 = call_lowered(@tvmgen_default_cutlass_main_0, (%x0, meta[relay.Constant][0], meta[relay.Constant][1]), metadata=meta[attributes][0]);
+          %2 = fn(%y_3_i0: Tensor[(600, 32, 64), float16], %y_3_i1: Tensor[(600, 32, 64), float16],
+                  Inline=1, Compiler="cublas", global_symbol="tvmgen_default_cublas_main_3", Primitive=1) -> Tensor[(600, 32, 32), float16] {
+            %6 = fn (%FunctionVar_0_01: Tensor[(600, 32, 64), float16], %FunctionVar_0_11: Tensor[(600, 32, 64), float16],
+                     PartitionedFromPattern="nn.batch_matmul_", Composite="cublas.batch_matmul") -> Tensor[(600, 32, 32), float16] {
+              nn.batch_matmul(%FunctionVar_0_01, %FunctionVar_0_11, out_dtype="float16", transpose_b=True)
+            };
+            %6(%y_3_i0, %y_3_i1)
+          };
+          %3 = %2(%x3, meta[relay.Constant][2]);
+          (%1, %3)
+        }
+        
+        def @tvmgen_default_cutlass_main_0(%y_0_i0: Tensor[(1600, 768), float16], %y_0_i1: Tensor[(2304, 768), float16], %y_0_i2: Tensor[(2304), float16],
+                  Extern=1) -> Tensor[(1600, 2304), float16] {
+          %4 = fn (%FunctionVar_0_0: Tensor[(1600, 768), float16], %FunctionVar_0_1: Tensor[(2304, 768), float16], %FunctionVar_0_2: Tensor[(2304), float16],
+                   PartitionedFromPattern="nn.dense_add_", Composite="cutlass.dense_bias") -> Tensor[(1600, 2304), float16] {
```

Comment:
This is unrelated question to this PR, but what are the difference between `PartitionedFromPattern`, `Composite` and `PrimFunc` at the perspective of lowering and fusion? They seem to serve similar purpose in a sense that they allow you lower a group of high-level ops to low-level.

Reply:
The "Primitive=1" attribute is a generic way to indicate 'this function represents a sub-graph which is to be lowered/compiled as a unit'. The additional "Compiler=foo" attributes indicates foo is the one doing the compiling, otherwise it's TVM. That kind makes sense.

The "PartitionedFromPattern" and "Composite" duplication seems unnecessary to me. Why not just use "PartitionedFromPattern"? The latter is always exactly equal to the former as per merge_composite.cc.

Reply:
I see. So if I understand correctly, `PartitionedFromPattern` is a more fine-grained way to specify lowering decision like a `Dense+Add` fusion. Meanwhile, `Primitive` can specify larger granularity of function that can contain multiple functions with different `PartitionedFromFusions`. 

Yeah, I agree that `Composite` seems to serve very similar role with `PartitionFromPattern` and would like to remove this potential duplication in the near future. 
It seemed bit tricky tho since some BYOCs seem to rely on `Composite` quite heavily.